doc/developer/design/00000000_template.md

```diff
@@ -1,42 +1,99 @@
 
-<!--
-// Enumerate potential goals that are explicitly out of scope for the project
-// ie. what could we do or what do we want to do in the future - but are not doing now
--->
+Explain the design as if it were part of Materialize and you were teaching the team about it.
+This can mean:
 
-## Description
+- Introduce new named concepts.
+- Explain the feature using examples that demonstrate product-level changes.
+- Explain how it builds on the current architecture.
+- Explain how engineers and users should think about this change, and how it influences how everyone uses the product.
+- If needed, talk though errors, backwards-compatibility, or migration strategies.
+- Discuss how this affects maintainability, or whether it introduces concepts that might be hard to change in the future.
 
-<!--
-// Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
-// If applicable, be sure to call out any new testing/validation that will be required
--->
+# Reference explanation
+[reference-explanation]: #reference-explanation
 
-## Alternatives
+Focus on the implementation of the feature.
+This is the technical part of the design.
 
-<!--
-// Similar to the Description section. List of alternative approaches considered, pros/cons or why they were not chosen
--->
+- Is it reasonably clear how the feature is implemented?
+- What dependencies does the feature have and introduce?
+- Focus on corner cases.
+- How can we test the feature and protect against regressions?
 
-## Open questions
+# Rollout
+[rollout]: #rollout
 
-<!--
-// Anything currently unanswered that needs specific focus. This section may be expanded during the doc meeting as
-// other unknowns are pointed out.
-// These questions may be technical, product, or anything in-between.
--->
+Describe what steps are necessary to enable this feature for users.
+How do we validate that the feature performs as expected? What monitoring and observability does it require?
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+Testability and explainability are top-tier design concerns!
+Describe how you will test and roll out the implementation.
+When the deliverable is a refactoring, the existing tests may be sufficient.
+When the deliverable is a new feature, new tests are imperative.
+
+Describe what metrics can be used to monitor and observe the feature.
+What information do we need to expose internally, and what information is interesting to the user?
+How do we expose the information?
+
+Basic guidelines:
+
+* Nearly every feature requires either Rust unit tests, sqllogictest tests, or testdrive tests.
+* Features that interact with Kubernetes additionally need a cloudtest test.
+* Features that interact with external systems additionally should be tested manually in a staging environment.
+* Features or changes to performance-critical parts of the system should be load tested.
+
```

Comment:
* Features that have upgrade implications or introduce migrations need to have either a legacy upgrade test or a Platform Check-based test
* Features that deal with the startup, the shutdown or the restart of any component of the system (components include environmentd, clusterd, CockroachDB, Kafka) need a mzcompose-based test